### PR TITLE
Add support for bigcode/starcoder ONNX export

### PR DESCRIPTION
**Pull request by xenova**
# What does this PR do?
When trying to run https://github.com/huggingface/optimum/tree/add-bigcode-support, I get the following error:

```
ValueError: Required inputs (['past_key_values.0.key_value', 'past_key_values.1.key_value', 'past_key_values.2.key_value', 'past_key_values.3.key_value', 'past_key_values.4.key_value', 'past_key_values.5.key_value', 'past_key_values.6.key_value', 'past_key_values.7.key_value', 'past_key_values.8.key_value', 'past_key_values.9.key_value', 'past_key_values.10.key_value', 'past_key_values.11.key_value', 'past_key_values.12.key_value', 'past_key_values.13.key_value', 'past_key_values.14.key_value', 'past_key_values.15.key_value', 'past_key_values.16.key_value', 'past_key_values.17.key_value', 'past_key_values.18.key_value', 'past_key_values.19.key_value', 'past_key_values.20.key_value', 'past_key_values.21.key_value', 'past_key_values.22.key_value', 'past_key_values.23.key_value']) are missing from input feed (['input_ids', 'attention_mask', 'past_key_values.0.key', 'past_key_values.0.value', 'past_key_values.1.key', 'past_key_values.1.value', 'past_key_values.2.key', 'past_key_values.2.value', 'past_key_values.3.key', 'past_key_values.3.value', 'past_key_values.4.key', 'past_key_values.4.value', 'past_key_values.5.key', 'past_key_values.5.value', 'past_key_values.6.key', 'past_key_values.6.value', 'past_key_values.7.key', 'past_key_values.7.value', 'past_key_values.8.key', 'past_key_values.8.value', 'past_key_values.9.key', 'past_key_values.9.value', 'past_key_values.10.key', 'past_key_values.10.value', 'past_key_values.11.key', 'past_key_values.11.value', 'past_key_values.12.key', 'past_key_values.12.value', 'past_key_values.13.key', 'past_key_values.13.value', 'past_key_values.14.key', 'past_key_values.14.value', 'past_key_values.15.key', 'past_key_values.15.value', 'past_key_values.16.key', 'past_key_values.16.value', 'past_key_values.17.key', 'past_key_values.17.value', 'past_key_values.18.key', 'past_key_values.18.value', 'past_key_values.19.key', 'past_key_values.19.value', 'past_key_values.20.key', 'past_key_values.20.value', 'past_key_values.21.key', 'past_key_values.21.value', 'past_key_values.22.key', 'past_key_values.22.value', 'past_key_values.23.key', 'past_key_values.23.value']).
```

so this branch/PR is just to do testing (and maybe ending up fixing it).

Edit: Following on from @JingyaHuang's [PR](https://github.com/huggingface/optimum/tree/add-bigcode-support), I added some logic to fix the exports and get it working for Transformers.js. I've tested and it seems to be working correctly ([demo](https://huggingface.co/spaces/Xenova/ai-code-playground)) 👍 

coding-demo-3

TODOs:
- [ ] At first, I was a little confused as to why the past/present key values were set to have the suffix `key_value` instead of `.key` and `.value` separately, but I assume this has something to do with the addition of multi-query attention. So, just want to make 100% sure about this.
- [ ] Documentation? 👀 
- [ ] Tests? 👀 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/exporters/onnx/base.py [hunk 1]
```diff
@@ -596,7 +596,8 @@ def generate_dummy_inputs(self, framework: str = "pt", **kwargs):
             and self.use_cache_branch is not False
             and "attention_mask" in dummy_inputs
         ):
-            past_length = dummy_inputs["past_key_values"][0][0].shape[2]
+            past_length = dummy_inputs["past_key_values"][0][0].shape[-2]
```
Comment (michaelbenayoun) [inline optimum/exporters/onnx/base.py:599]: Why?

File: optimum/exporters/onnx/model_configs.py [hunk 2]
```diff
@@ -268,6 +268,47 @@ def add_past_key_values(self, inputs_or_outputs: Dict[str, Dict[int, str]], dire
...
+        )
+        return [self.random_float_tensor(past_key_value_shape, framework=framework) for _ in range(self.num_layers)]
+
+
+class GPTBigCodeOnnxConfig(TextDecoderOnnxConfig):
+    DUMMY_INPUT_GENERATOR_CLASSES = (
+        GPTBigCodeDummyPastKeyValuesGenerator,
+    ) + TextDecoderOnnxConfig.DUMMY_INPUT_GENERATOR_CLASSES
+    DUMMY_PKV_GENERATOR_CLASS = GPTBigCodeDummyPastKeyValuesGenerator
+    NORMALIZED_CONFIG_CLASS = NormalizedTextConfig.with_args(num_layers="n_layer", num_attention_heads="n_head")
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/exporters/onnx/model_configs.py:286]: Hi @michaelbenayoun, thanks for the review. I will address your comments in my PR [here](https://github.com/huggingface/optimum/pull/1042)

File: optimum/exporters/onnx/base.py (hunk 1 above)
Comment (JingyaHuang) [reply optimum/exporters/onnx/base.py:599]: With multi-query attention, there is one dimension less (num_heads), so I used reversed indexing.

Comment (michaelbenayoun) [reply optimum/exporters/onnx/base.py:599]: Alright, as long as it works for everyone.

Comment (JingyaHuang) [reply optimum/exporters/onnx/base.py:599]: Yes it is, seq_len is in the same position for every model if we count from a reversed order.

File: optimum/exporters/onnx/model_configs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/exporters/onnx/model_configs.py:286]: There is already gpt bigcode normalized config thanks to the [PR from Intel](https://github.com/huggingface/optimum/commit/53e09fe42ef774d6f8d2c5cfd8bd19e1eee84c0d), I will use it directly then.

**Comment by JingyaHuang** [comment]
Hi @xenova, 

For your first question in todo, yes, as in gpt-bigcode when using multi-query attn, the [key and value are concatenated](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py#L244-L246) together as output, I am not using the regular suffix, otherwise there will be two times of output for the traced model and won't match.

File: optimum/exporters/onnx/model_configs.py [hunk 3]
```diff
@@ -268,6 +268,47 @@ def add_past_key_values(self, inputs_or_outputs: Dict[str, Dict[int, str]], dire
...
+            name = "present"
+
+        for i in range(self._normalized_config.num_layers):
+            # No dim for `n_head` when using multi-query attention
+            inputs_or_outputs[f"{name}.{i}.key_value"] = {
+                0: "batch_size",
+                1: decoder_sequence_name,
+            }
+
+    def flatten_past_key_values(self, flattened_output, name, idx, t):
```
[the remark this answers is not in the document]
Comment (xenova) [reply optimum/exporters/onnx/model_configs.py:309]: Sweet! Go for it 😄 ... yeah it took me an hour or two of scrolling through the code to figure out what was going wrong, and I figured that would be the simplest place to put it (as I'm sure there are other ways)

[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/exporters/onnx/model_configs.py:309]: @xenova Yeah, for the mqa case, there is no need to flatten as we leverage key and value as a whole. It's really nice to use the method you overridden. It's better to put a dummy flatten method under bigcode-specific config than create an ugly conditional statement in the general validation method(the first thought that came into my mind), thanks a lot! 

My PR for supporting gpt bigcode export + ort inference is ready, also tagged you for visibility, thanks again !

[the remark this answers is not in the document]
Comment (xenova) [reply optimum/exporters/onnx/model_configs.py:309]: 👍  I'll test your PR with transformers.js tomorrow :)

**Pull request closed without merging**